File: bjst/stats.py
class Statistics:
    """Statistics tracking for blackjack strategy training sessions.

    This class tracks performance metrics during training sessions, including:
    - Overall accuracy (correct answers / total attempts)
    - Accuracy by hand type (hard totals, soft totals, pairs)
    - Accuracy by dealer strength (weak, medium, strong dealer cards)

    Dealer strength categories:
    - Weak: 4, 5, 6 (dealer bust cards)
    - Medium: 2, 3, 7, 8 (moderate dealer cards)
    - Strong: 9, 10, A (strong dealer cards)

    The statistics are maintained for the current session and can be displayed
    to show the user's progress and identify areas for improvement.
    """

    def __init__(self):
        self.session_data = {
            'total_attempts': 0,
            'correct_answers': 0,
            'by_category': {
                'hard': {'correct': 0, 'total': 0},
                'soft': {'correct': 0, 'total': 0},
                'pair': {'correct': 0, 'total': 0}
            },
            'by_dealer_strength': {
                'weak': {'correct': 0, 'total': 0},
                'medium': {'correct': 0, 'total': 0},
                'strong': {'correct': 0, 'total': 0}
            }
        }
        self.overall_accuracy = {}

    def record_attempt(self, hand_type, dealer_strength, correct):
        self.session_data['total_attempts'] += 1
        if correct:
            self.session_data['correct_answers'] += 1

        # Record by hand type
        if hand_type in self.session_data['by_category']:
            self.session_data['by_category'][hand_type]['total'] += 1
            if correct:
                self.session_data['by_category'][hand_type]['correct'] += 1

        # Record by dealer strength
        if dealer_strength in self.session_data['by_dealer_strength']:
            self.session_data['by_dealer_strength'][dealer_strength]['total'] += 1
            if correct:
                self.session_data['by_dealer_strength'][dealer_strength]['correct'] += 1

    def get_category_accuracy(self, category):
        data = self.session_data['by_category'].get(
            category, {'correct': 0, 'total': 0})
        if data['total'] == 0:
            return 0.0
        return (data['correct'] / data['total']) * 100

    def get_dealer_strength_accuracy(self, strength):
        data = self.session_data['by_dealer_strength'].get(
            strength, {'correct': 0, 'total': 0})
        if data['total'] == 0:
            return 0.0
        return (data['correct'] / data['total']) * 100

    def get_session_accuracy(self):
        if self.session_data['total_attempts'] == 0:
            return 0.0
        return (self.session_data['correct_answers'] /
                self.session_data['total_attempts']) * 100

    def display_progress(self):
        print("\n" + "=" * 50)
        print("SESSION STATISTICS")
        print("=" * 50)

        total = self.session_data['total_attempts']
        correct = self.session_data['correct_answers']

        if total == 0:
            print("No practice attempts yet this session.")
            return

        print(
            f"Overall: {correct}/{total} ({self.get_session_accuracy():.1f}%)")

        print("\nBy Hand Type:")
        for hand_type in ['hard', 'soft', 'pair']:
            data = self.session_data['by_category'][hand_type]
            if data['total'] > 0:
                accuracy = (data['correct'] / data['total']) * 100
                print(f"  {hand_type.capitalize()}: {data['correct']}/{data['total']} "
                      f"({accuracy:.1f}%)")

        print("\nBy Dealer Strength:")
        for strength in ['weak', 'medium', 'strong']:
            data = self.session_data['by_dealer_strength'][strength]
            if data['total'] > 0:
                accuracy = (data['correct'] / data['total']) * 100
                print(f"  {strength.capitalize()}: {data['correct']}/{data['total']} "
                      f"({accuracy:.1f}%)")

        input("\nPress Enter to continue...")

    def reset_session(self):
        self.session_data = {
            'total_attempts': 0,
            'correct_answers': 0,
            'by_category': {
                'hard': {'correct': 0, 'total': 0},
                'soft': {'correct': 0, 'total': 0},
                'pair': {'correct': 0, 'total': 0}
            },
            'by_dealer_strength': {
                'weak': {'correct': 0, 'total': 0},
                'medium': {'correct': 0, 'total': 0},
                'strong': {'correct': 0, 'total': 0}
            }
        }

    def get_dealer_strength(self, dealer_card):
        if dealer_card in [4, 5, 6]:
            return 'weak'
        if dealer_card in [2, 3, 7, 8]:
            return 'medium'
        # 9, 10, 11 (Ace)
        return 'strong'

File: go/internal/stats/stats.go
// Package stats provides statistics tracking for blackjack strategy training sessions.
//
// This package tracks performance metrics during training sessions, including:
// - Overall accuracy (correct answers / total attempts)
// - Accuracy by hand type (hard totals, soft totals, pairs)
// - Accuracy by dealer strength (weak, medium, strong dealer cards)
//
// Dealer strength categories:
// - Weak: 4, 5, 6 (dealer bust cards)
// - Medium: 2, 3, 7, 8 (moderate dealer cards)
// - Strong: 9, 10, A (strong dealer cards)
//
// The statistics are maintained for the current session and can be displayed
// to show the user's progress and identify areas for improvement.
package stats

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// CategoryData tracks correct and total attempts for a category.
type CategoryData struct {
	Correct int
	Total   int
}

// Statistics tracks performance metrics for training sessions.
type Statistics struct {
	totalAttempts    int
	correctAnswers   int
	byCategory       map[string]*CategoryData
	byDealerStrength map[string]*CategoryData
}

// New creates a new statistics tracker.
func New() *Statistics {
	stats := &Statistics{
		totalAttempts:    0,
		correctAnswers:   0,
		byCategory:       make(map[string]*CategoryData),
		byDealerStrength: make(map[string]*CategoryData),
	}

	// Initialize category tracking
	stats.byCategory["hard"] = &CategoryData{}
	stats.byCategory["soft"] = &CategoryData{}
	stats.byCategory["pair"] = &CategoryData{}

	// Initialize dealer strength tracking
	stats.byDealerStrength["weak"] = &CategoryData{}
	stats.byDealerStrength["medium"] = &CategoryData{}
	stats.byDealerStrength["strong"] = &CategoryData{}

	return stats
}

// RecordAttempt records an attempt in the training session.
func (s *Statistics) RecordAttempt(handType, dealerStrength string, correct bool) {
	s.totalAttempts++
	if correct {
		s.correctAnswers++
	}

	// Record by hand type
	if category, exists := s.byCategory[handType]; exists {
		category.Total++
		if correct {
			category.Correct++
		}
	}

	// Record by dealer strength
	if strength, exists := s.byDealerStrength[dealerStrength]; exists {
		strength.Total++
		if correct {
			strength.Correct++
		}
	}
}

// GetCategoryAccuracy returns accuracy percentage for a specific category.
func (s *Statistics) GetCategoryAccuracy(category string) float64 {
	if data, exists := s.byCategory[category]; exists && data.Total > 0 {
		return (float64(data.Correct) / float64(data.Total)) * 100.0
	}
	return 0.0
}

// GetDealerStrengthAccuracy returns accuracy percentage for a dealer strength category.
func (s *Statistics) GetDealerStrengthAccuracy(strength string) float64 {
	if data, exists := s.byDealerStrength[strength]; exists && data.Total > 0 {
		return (float64(data.Correct) / float64(data.Total)) * 100.0
	}
	return 0.0
}

// GetSessionAccuracy returns overall session accuracy percentage.
func (s *Statistics) GetSessionAccuracy() float64 {
	if s.totalAttempts == 0 {
		return 0.0
	}
	return (float64(s.correctAnswers) / float64(s.totalAttempts)) * 100.0
}

// DisplayProgress displays progress statistics to the console.
func (s *Statistics) DisplayProgress() {
	fmt.Println("\n" + strings.Repeat("=", 50))
	fmt.Println("SESSION STATISTICS")
	fmt.Println(strings.Repeat("=", 50))

	if s.totalAttempts == 0 {
		fmt.Println("No practice attempts yet this session.")
		fmt.Print("\nPress Enter to continue...")
		bufio.NewReader(os.Stdin).ReadString('\n')
		return
	}

	fmt.Printf("Overall: %d/%d (%.1f%%)\n",
		s.correctAnswers, s.totalAttempts, s.GetSessionAccuracy())

	fmt.Println("\nBy Hand Type:")
	for _, handType := range []string{"hard", "soft", "pair"} {
		if data, exists := s.byCategory[handType]; exists && data.Total > 0 {
			accuracy := (float64(data.Correct) / float64(data.Total)) * 100.0
			capitalized := strings.Title(handType)
			fmt.Printf("  %s: %d/%d (%.1f%%)\n", capitalized, data.Correct, data.Total, accuracy)
		}
	}

	fmt.Println("\nBy Dealer Strength:")
	for _, strength := range []string{"weak", "medium", "strong"} {
		if data, exists := s.byDealerStrength[strength]; exists && data.Total > 0 {
			accuracy := (float64(data.Correct) / float64(data.Total)) * 100.0
			capitalized := strings.Title(strength)
			fmt.Printf("  %s: %d/%d (%.1f%%)\n", capitalized, data.Correct, data.Total, accuracy)
		}
	}

	fmt.Print("\nPress Enter to continue...")
	bufio.NewReader(os.Stdin).ReadString('\n')
}

// ResetSession resets session statistics.
func (s *Statistics) ResetSession() {
	s.totalAttempts = 0
	s.correctAnswers = 0

	for _, category := range s.byCategory {
		category.Correct = 0
		category.Total = 0
	}

	for _, strength := range s.byDealerStrength {
		strength.Correct = 0
		strength.Total = 0
	}
}

// GetDealerStrength determines dealer strength from dealer card.
func (s *Statistics) GetDealerStrength(dealerCard int) string {
	switch dealerCard {
	case 4, 5, 6:
		return "weak"
	case 2, 3, 7, 8:
		return "medium"
	default: // 9, 10, 11 (Ace)
		return "strong"
	}
}
